[Merged by Bors] - feat(number_theory/arithmetic_function): The moebius function is multiplicative

khwilson opened the pull request:
A fundamental property of the moebius function is that it is
multiplicative, which allows many facts about Euler products to be
expressed



---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

khwilson committed:
feat(number_theory/arithmetic_function): The moebius function is multiplicative

A fundamental property of the moebius function is that it is
multiplicative, which allows many facts about Euler products to be
expressed

khwilson committed:
Linting

khwilson committed:
Small fix during copy paste

khwilson committed:
One more try for the debugger

eric-wieser commented on src/algebra/squarefree.lean:
```diff
@@ -49,6 +49,14 @@ begin
 end
 
+lemma ne_zero_of_squarefree [monoid_with_zero R] [nontrivial R] {m : R}
+(hm : squarefree (m : R)) : m ≠ 0 :=
```
```suggestion
lemma squarefree.ne_zero [monoid_with_zero R] [nontrivial R] {m : R}
  (hm : squarefree (m : R)) : m ≠ 0 :=
```

khwilson replied:
Oh very nice!

eric-wieser commented on src/algebra/squarefree.lean:
```diff
@@ -49,6 +49,14 @@ begin
+  by_contradiction H,
+  rw H at hm,
+  exact not_squarefree_zero hm,
+end
```
Could golf to
```suggestion
λ h, not_squarefree_zero $ (h ▸ hm : squarefree (0 : R))
```
or the middle ground
```suggestion
begin
  rintro rfl,
  exact not_squarefree_zero hm,
end
```

eric-wieser commented on src/algebra/squarefree.lean:
```diff
@@ -422,6 +430,58 @@ lemma squarefree_iff_prime_sq_not_dvd (n : ℕ) :
+lemma squarefree_mul {m n : ℕ} (hmn : m.coprime n) :
+squarefree m ∧ squarefree n ↔ squarefree (m * n) :=
+begin
+  split,
```
Mathlib style is to use `{}` after split to make it clear which tactics are operating on which direction; so in this case, the proof would proceed as:
```lean
split,
{ rintros ⟨hm, hn⟩,
  -- solve the forward direction here
  sorry },
{ intros hmn_sq,
  -- solve the backwards direction here
  sorry },
```

ericrbg replied:
also, you can use `refine ⟨_, λ a, _⟩` to do the `split` and `intros` at the same time (sadly this doesn't work with `rintro`)

khwilson replied:
Thanks for that! Neat trick. Though bhavik's suggestion obviates all of this :-)

eric-wieser commented on src/number_theory/arithmetic_function.lean:
```diff
@@ -732,6 +732,25 @@ begin
+      simp [this, hn_sq], }, },
+  have : ¬ squarefree (m * n),
+    by_contradiction H,
+    exact hm_sq ((squarefree_mul hmn).mpr H).left,
```
Similarly to above; mathlib style is to use `{}` not indents, as `{}` can't lie about where the goals are:
```suggestion
  have : ¬ squarefree (m * n),
  { by_contradiction H,
    exact hm_sq ((squarefree_mul hmn).mpr H).left },
```

khwilson replied:
Thanks! Trying to learn the style. Is there a linter available?

ericrbg replied:
all the linters that exist run automatically; you can run them by hand by putting `#lint` on a blank line. There is a style guide, though: https://leanprover-community.github.io/contribute/style.html

khwilson replied:
Thanks so much!

khwilson committed:
Update src/algebra/squarefree.lean

I'm new enough that I only understand this one, so let's go with this :-)

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

khwilson committed:
Update src/algebra/squarefree.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

b-mehta commented on src/algebra/squarefree.lean:
```diff
@@ -422,6 +429,58 @@ lemma squarefree_iff_prime_sq_not_dvd (n : ℕ) :
+    calc x * x ∣ n : hx
+      ... ∣ m * n : dvd_mul_left n m,
+  exact hxx (hmn_sq this),
+end
```
```suggestion
lemma squarefree_mul {m n : ℕ} (hmn : m.coprime n) :
  squarefree m ∧ squarefree n ↔ squarefree (m * n) :=
begin
  simp only [squarefree_iff_prime_squarefree, ←sq, ←forall_and_distrib],
  refine ball_congr (λ p hp, _),
  simp only [hmn.is_prime_pow_dvd_mul (hp.is_prime_pow.pow two_ne_zero), not_or_distrib],
end
```
This proof can be simplified a little

khwilson replied:
Oh, a much simpler proof! Learning new techniques (`ball_congr` !) every day! Thank you :-)

b-mehta commented on src/number_theory/arithmetic_function.lean:
```diff
@@ -732,6 +732,25 @@ begin
 end
 
+lemma is_multiplicative_moebius : is_multiplicative μ :=
+⟨by simp, begin
```
With an extra lemma about multiplicative functions in general, this simplifies too:
```
lemma iff_ne_zero [monoid_with_zero R] {f : arithmetic_function R} :
  is_multiplicative f ↔
    f 1 = 1 ∧ (∀ {m n : ℕ}, m ≠ 0 → n ≠ 0 → m.coprime n → f (m * n) = f m * f n) :=
begin
  refine and_congr_right' (forall₂_congr (λ m n, ⟨λ h _ _, h, λ h hmn, _⟩)),
  rcases eq_or_ne m 0 with rfl | hm,
  { simp },
  rcases eq_or_ne n 0 with rfl | hn,
  { simp },
  exact h hm hn hmn,
end
```
then
```
lemma is_multiplicative_moebius : is_multiplicative μ :=
begin
  rw is_multiplicative.iff_ne_zero,
  refine ⟨by simp, λ n m hn hm hnm, _⟩,
  simp only [moebius, zero_hom.coe_mk, ←squarefree_mul hnm, ite_and, card_factors_mul hn hm],
  rw [pow_add, ite_mul_zero_right, ite_mul_zero_left],
end
```

khwilson replied:
Thanks! Requires adding an import, but I assume that's fine. :-)

b-mehta replied:
Oops I should've mentioned that, sorry!

khwilson committed:
Update src/algebra/squarefree.lean

Vastly simplified proof of `squarefree_mul`

Co-authored-by: Bhavik Mehta <bhavikmehta8@gmail.com>

khwilson committed:
Simplifying proof that Moebius is multiplicative

Simplification by @b-mehta. Adds `iff_ne_zero` for proving a function is
multiplicative.

khwilson committed:
Add import algebra.is_prime_pow to algebra.squarefree

This allows for the proof in the previous commit to compile

khwilson committed:
Merge remote-tracking branch 'origin/master' into moebius_multiplicative

eric-wieser commented on src/algebra/squarefree.lean:
```diff
@@ -450,6 +458,14 @@ lemma squarefree_iff_prime_sq_not_dvd (n : ℕ) :
   ⟨2, (irreducible_iff_nat_prime _).2 prime_two⟩
 
+lemma squarefree_mul {m n : ℕ} (hmn : m.coprime n) :
+  squarefree m ∧ squarefree n ↔ squarefree (m * n) :=
```
Do both directions require hmn?

khwilson replied:
No, though the thought was here to mimic the idea of a "multiplicative function." I can create a `squarefree_mul'` that looks like:

```lean
/- Restatement of one direction of `squarefree_mul` without the coprimality assumption -/
lemma squarefree_mul' {m n : ℕ} : squarefree (m * n) → squarefree m ∧  squarefree n := ...
```

b-mehta replied:
I think this would be a good idea, `squarefree.of_mul_left` for `squarefree (m * n) -> squarefree m` (perhaps Eric will have a better name)

khwilson replied:
Added with the suggested names.

eric-wieser replied:
```suggestion
  squarefree (m * n) ↔ squarefree m ∧ squarefree n :=
```
Since we usually name iffs by their LHS

khwilson committed:
Add squarefree.of_mul_(left|right)

These are auxiliary lemmas that don't require coprimality for one
direction of (squarefree m ∧ squarefree n ↔ squarefree (m * n))

khwilson committed:
Update src/algebra/squarefree.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

khwilson committed:
Fix downstream after lemma rewrite

khwilson committed:
Remove extra space

khwilson committed:
Fix comment syntax

eric-wieser commented on src/algebra/squarefree.lean:
```diff
@@ -450,6 +464,16 @@ lemma squarefree_iff_prime_sq_not_dvd (n : ℕ) :
 squarefree_iff_irreducible_sq_not_dvd_of_exists_irreducible
   ⟨2, (irreducible_iff_nat_prime _).2 prime_two⟩
 
+/-- `squarefree` is multiplicative. Note that the ← direction does not require `hmn`.
```
```suggestion
/-- `squarefree` is multiplicative. Note that the → direction does not require `hmn` and generalizes to
arbitrary commutative monoids.
```

khwilson replied:
Good catch!

khwilson committed:
Update src/algebra/squarefree.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

khwilson committed:
Fix linting

jcommelin commented on src/algebra/squarefree.lean:
```diff
@@ -450,6 +464,17 @@ lemma squarefree_iff_prime_sq_not_dvd (n : ℕ) :
   ⟨2, (irreducible_iff_nat_prime _).2 prime_two⟩
 
+/-- `squarefree` is multiplicative. Note that the → direction does not require `hmn`
+and generalizes to arbitrary commutative monoids. See `squarefree.of_mul_left` and
```
It should be true in arbitrary `unique_factorization_monoid`s, right?

jcommelin reviewed (approved):
Thanks :tada:

bors merge
